### ISSUE TITLE
Code review - Run0rD1e

### ISSUE THREAD

**Issue by dialexo** [Zeavee/Run0rD1e#181]
# Code review

This is a code review for the code of the whole project on master branch at commit https://github.com/Zeavee/Run0rD1e/tree/3ed29523faa2f3c31fea2e58ac5c6ce1ec812dbb

Of course, we don't expect you to change all of your code (some parts of the review can be subjective or arguable anyway), what’s important is that you understand why better alternatives may exist.

Please ask us if you have any question about the code review, whether it is by writing a comment below or by sending us an email or meeting us.

**Important : if you try to fix these, don't break your app. We prefer that the app works even if you don't manage to fix anything. It may be too hard to rework many of the issues below, as it would require to rewrite the whole architecture.**

### Packages structure 

The depth of your package tree is almost always 1. We encourage you to create sub packages where needed. It’s a good thing that you separated your code using packages, however right now the package structure makes it difficult to understand where the code is.

You seem to split the packages by **component** rather than by **functionality** (instead of having a database package, a activities package…, you have a item package, a leaderboard package…). This is common in microservices architecture, but Android doesn’t seem very well suited for that. It’s probably more convenient to have all your DAOs in the same db package, for example.

> As an example of this problem, we found that `artificial_intelligence` contains lots of things unrelated to artificial intelligence (i.e. `Boundable`, `CartesianPoint`, `Movable`, `Localizable`, `GenPoint` and its implementations). You should probably reorder your packages using the `entities` package as a root for all this. Another example is that you have a `db` package and a `database` package.

Overall, a lot of classes implement / extend classes that are in other packages. As part of your refactoring, you should probably move these classes and interfaces around, so that interfaces and most (with some exceptions, such as `Callbacks`) of their implementations are in the same package. We suggest you a package hierarchy looking like that:

```
* database/
   * firebase/
   * room/
   * repositories/
* entities/
* items/
* game/
* geometry/
* ui/
   * social/
   * leaderboard/
   * map/
* ...
```

Should you chose to keep a component-oriented structure, you should consider each of these packages as a “mini app”: the package should itself be split in multiple subpackages (`ui`, `database`, …).

### Documentation

We noticed that a lot of classes and method lack documentation. 

This problem is especially annoying for interfaces, as they _declare_ the functionality. There is no way for the developer that will implement it nor the one that will use it to fully understand how the implementation should behave. It makes the code way harder to read: to understand what an interface does, one is forced to find an implementation and read its code.

We therefore had a lot trouble to understand your code and evaluate its implementation as without formal specification of what you’re trying to accomplish, it is very difficult to verify if it adheres to it.

**All public classes, methods and fields must be documented.** Use JavaDoc conventions: a top-level block for general description and metadata, and per-method header with annotations (`@param`, `@return`, `@throws`, ...). You can omit JavaDoc on overriden methods if they are themselves documented.

### Code style

https://en.wikipedia.org/wiki/Naming_convention_(programming)#Java

We noticed inconsistent code style across your codebase. For example, for brackets positioning, some of you do this:

```java
if (stuff) {
  // do things
}
```

while others do that:

```java
if (stuff)
{
 // do things
}
``` 

The first one is considered to be the “java best practice”, but you’re free to pick the second one as long as you all follow the same practice throughout your codebase.

**You should choose a code style and stick to it.**. The default one in Android Studio is already very good, but if you want you can also pick for example the Google Java style (see here: https://google.github.io/styleguide/javaguide.html). It also comes with a Gradle plugin that can reformat all your code with a single command (see here: https://medium.com/@alexprut/integrate-google-java-style-guide-in-a-java-project-567abb6d7987).

In anycase, you should program your codestyle (or use the default) in your IDE and use CTRL-ALT-L shortcut to reformat your files and optimise imports.

### Code quality & performance

You should pay attention to the lints and warnings that Android Studio highlights for you (`Menu bar > Analyze > Inspect Code…`). 

We also noticed a few unused classes in your code:
* `MovementType` (enum) is unused
* `ItemType` (enum) is unused

There are also unused fields and methods in some of your classes. Ideally, there should be no such thing on the master branch.

Regarding code you no longer want to use, just commenting it is “okay” as long as it stays on your computer or in a branch. 

Code that is unused or commented out makes reading the code unneededly harder. There should never be such things on the master branch, and your PR reviews should get rid of them.

We found several hardcoded strings and magic values in your code, including in your XML layout files (use the `strings.xml` and `values.xml` files).

Make sure to always be precise: your variable names are not always meaningful. Semantics are crucial as they are the only way for you to convey your intentions. Another example is class names: you should always keep the underlying type as suffix, and the name should match the class functionality.

### Class hierarchy

We find that your class hierarchy is often labyrinthic. This particularly applies to the _entities_, _positions_ and _displayables_ but remains valid for the rest. We suggest that you try to draw your class hierarchy on a piece of paper, including the relationships between the classes, to see what is wrong with it.
To come up with a better hierarchy, we suggest that you also use some paper and try to draw it before coding. You can user your _Items_ hierarchy as an example of overall good class hierarchy. 

For more specific examples :
* You have 3 different types of positions, and one of them (GeoPoint) is not an instance of `GenPoint`. This forces you to duplicate some interfaces that provide similar functionality but use different types of positioning. This also forces a class implementing both interfaces to use two positions at the same time (see the `Player` class).
You have two options here (we believe the second one would be easier for you to use):
   * Make GeoPoint an instance of GenPoint
   * Choose a single coordinates system (Geodesic or Cartesian) and stick to it: all positions should be expressed in this system, and if you get positions in other systems from external sources (i.e. GPS) you convert them immediately before feeding them to your classes.
   * `Displayable` defines a `getEntityType` method. This method is also defined by some items, independently. This clearly means that the `getEntityType` method should be declared by some other interface (probably `Entity`).

At the opposite, we also found some places in which your abstractions were over-engineered, with some layers of abstraction that didn’t bring any actual abstraction and only made the code harder to read. In general, try to write the code that is the most precise, and [short as possible](https://en.wikipedia.org/wiki/KISS_principle). You will always face trade-offs, and sometimes you should prioritize cognitive load against over-abstraction.

### MainActivity.java

* What is the goal of your `MainActivity`? Is it the entry point that initializes all components of your app and recursively build them using dependency injection? Or is it just a main screen, in which case you will need some other class that will hold all the required state?
* No JavaDoc: top-level block and methods (excluding overrides)
* Line 22: the class holds an unset reference to `game`. Also on line 25, `startGame` is unused. Also see comments for [#Game.java].
* Why is the `AuthenticationController` public (line 37) ? As you instantiate it here, you should be able to pass it to other components that use it through dependency injection, thus the field should be private (eventually final).
* Line 65: `new DatabaseHelper(this).deleteAllUsers()` : the class shouldn’t be initialized in place: this makes it impossible to test it independently.
* Line 66: `LoginFormActivity.loggedUser = null` : this is a bad practice that you set a field of another activity: you have no guarantee that the Activity actually exists (you have no control over the lifecycle of activities, the OS does), and it makes your code interdependent. What is `loggedUser` for? It should not be a global variable (which it is since it is public static) and it is unset anyway (see remarks on [#LoginFormActivity.java]).

### artificial_intelligence/

This package contains a lot of classes that often are very lightly related. 

#### Behaviour.java

This enum is only used in Enemy.java. Shouldn’t then this enum be part privately of the Enemy class? Also, document what the enum models.

#### Boundable, Localizable, Movable, Updatable :

The name of these interfaces doesn’t make it clear what they do:
 
* `Boundable` looks like something that can be constrained by bounds, rather than the bounds themselves (could be named `Area` or `Bounds`)
* `Localizable` → `Positioned` (in english, Localized means translated, not positioned)
* `Updatable` → not clear what an update should do. What is updated? What does it mean to update? It looks like a [game tick](https://gamedev.stackexchange.com/questions/81608/what-is-a-tick-in-the-context-of-game-development) to us, maybe it should be called `Tickable`
* Movable is fine.

You should also add Javadoc to further explain what the method is supposed to do. For example, `Movable.move()` doesn’t take arguments: you should therefore explain what its behaviour is supposed to be. `Updatable.update()` should also be clearly explained.

As a side note, the way `Movable` is defined now forces you to define most of the behaviour of the entities inside this method and the entities themselves. This is mostly fine, but you could also have created an interface extending `Positioned` which exposes a single method to set the position. **You don’t need to change that as your implementations work and are correct. It seems you already did that partially in the `MovingEntity` class.**

#### CartesianPoint, GenPoint, PolarPoint

The abstraction seems a bit over-engineered and artificial: the fact that your two fields are named “arg1” and “arg2” shows that. From our point of view, what your `GenPoint` class really exposes is the `toCartesian` and `toPolar` methods. It means that GenPoint could actually be an interface called Point only exposing these methods. Maybe you would also want to abstractly define `add`, `normalize`, `distanceFrom`, … in your `Point` interface. This interface would also contain some JavaDoc describing the way it works! Also you should document the concrete implementations.

`CartesianPoint` could expose the getX() and getY() methods as substitutes for `getArg1()` and `getArg2()` if you need them (similar considerations for `PolarPoint`).

Why is there an empty constructor? A data class should never have an inconsistent state. Also, consider making the class immutable (all properties final and private) and returning copies should you modify it (for instance `normalize`) since a mathematical representation of a 2D position can never change.

What does it mean to multiply a 2d point? A vector and a point is not the same mathematical object: you can consider a Point to be a Vector but the reverse is not true.

#### Enemy.java

* Using a state machine for behaviour is a nice idea.
* No JavaDoc, it makes it extremely hard for us to guess the functionality and implementation of your classes.
* Why is there a reference to the list of all players? From a modelling perspective it doesn’t make sense, as an enemy entity does not “have” players. You could call the `PlayerManager` directly at line 134. `selectClosestPlayer()` could be a method of `PlayerManager` which takes a position as argument, it makes sense that the manager is more informed about all the players than an enemy.

* Lines 13,14: `damage` and `dps` are misnamed: is `dps` the damage rate (instead of actual damage amount per second)? Is `damage` the quantity of damage done per hit? This is not semantically clear. 
* Shouldn’t `timeAttack` be named `attackTimeDelay` (and resp. `wanderingTimeDelay`) ? The Javadoc should make it clear what these fields contain (e.g. the number of frames before the next attack / wander). 
* You should not use magic values (`timeAttack = 30`, line 111). You should define a global constant FPS somewhere (make `GameThread.FPS` `public static final` ?) and use it everywhere you need that constant.
* `checkWaiting()` can be made private?

#### EnemyGenerator.java

* JavaDoc would be nice
* We don’t understand why it’s stateful: should it be an interface that simply describes what an enemy generator does? Or if it needs to store state, is it useful that your setters mutate it? Wouldn’t it be possible to instantiate a new immutable one with the constructor when you require different parameters?
* What does the `rule()` method do, does it simply generate a new position (naming is unclear) ? It is hard for us to say if your implementation is right or wrong since the absence of documentation make your code’s intentions unclear.
* From what we’ve seen you only use it for `RandomEnemyGenerator`, thus we think this interface is overengineered. Either you simply expose the abstract methods, or you simply remove it as the `RandomEnemyGenerator` stands for itself (except if you had planned to design more `EnemyGenerators`).
* Line 37: you may want to use `Collection.immutableList(this.enemies)` instead of a loop. Or just the copy constructor `new ArrayList<>(this.enemies)`

#### LocalBounds.java

The class should be immutable. 

#### Movement.java

The abstraction is clunky without documentation:

* If your goal was to represent a function which takes a position and return a new one given the parameters then it’s mostly fine
* Otherwise, why does a `Movement` have a position as property? Wouldn’t it make more sense for it to replace `nextPosition` with a method which takes a position (or a `Movable` object) as argument and returns a new position (resp. moves the `Movable`).

#### MovingArtificialEntity.java

* No JavaDoc
* `switchOnMouvement()`: this method is not well named, keep everything in English. Also shouldn’t it be named `bounce()`? It also seems strange to have such a concrete implementation of it in an abstract class (although it could a default later overridden, in which case it’s fine).
* Why isn’t the class abstract? You never instantiate it, and it would not make sense to do so. It forces you to implement defaults that are not sane (for instance `getEntityType`).
* Why is `getEntityType` overridden here, since it returns `EntityType.NONE`? Again, if it’s a default, it’s fine otherwise you should just leave it abstract for now (and define it in children classes).

#### PointConverter.java

* Nice method documentation. You should also add a class Javadoc block.

#### RandomEnemyGenerator.java

* No JavaDoc
* Unused `mapEnemiesToTiles` and you should comment the line after
* Lines 71, 72: you could avoid unnecessary type (un)boxing by using `float` instead of `Float`

#### RectangleBounds.java

* No JavaDoc
* The class should be immutable
* There shouldn’t be commented code in production code

### database/

This package should probably become `database/firebase` or similar.

#### FirestoreUserData.java

* No JavaDoc
* Should be named `FirestoreUserDataController` to keep the type suffix

#### UserDataController.java

* No JavaDoc
* This interface should be an agnostic to its implementation, but you have a function named `syncCloudFirebaseToRoom`. (Here, the name implies that the implementation will use firebase and room.)

### db/

This package should probably become `database/sql` or similar.

#### AppDatabase.java

* No class JavaDoc, we don’t understand the intentions
* `sRoomDatabaseCallback` and `addDelay`: is this testing code in production code? Do you need to empty the database at every app launch? Did you use this code https://github.com/googlecodelabs/android-room-with-a-view/blob/master/app/src/main/java/com/example/android/roomwordssample/WordRoomDatabase.java#L67? **Adding delay artificially is a very bad practice** as it completely blocks a thread. Removing this part will allow you to remove `mIsDatabaseCreated` as well (as we don’t see any particular purpose for this LiveData).

#### LeaderoardViewModel.java

Typo in the name, should be LeaderboardViewModel.java

#### EntityType.java

* You probably want an interface / abstract class `Entity` which defines an inner enum `Type` instead, from which you inherit all game entities. Otherwise, there is actually no abstraction as you already need to know the type of the item before being able to call `getEntityType`:

> Example : `HealthPack` defines the `getEntityType` method. It is not defined in `Item`, so it means that when you call `healthPack.getEntityType`, you already know that it’s a `HealthPack` since you could not call `getEntityType` on it otherwise.

#### MovingEntity.java

* It seems weird to implement `aoeRadius` here as it seems that the Moving class should only implement the logic to move an entity.

#### Player.java

* No JavaDoc
* The `Inventory` is a nice abstraction
* It would make sense to have a class converter that separates the concerns of the Player data model and the Firebase storage of the Player. Indeed, you are forced to use public modifiers, annotations which make your implementation dependent on external dependencies and vulnerable. For instance, you use a `Timestamp` proprietary type from Firebase. Also, you now have public fields AND redundant getters.
* The fact that you implement both `Displayable` (via `MovingEntity`) and `Localizable` forces you to maintain the position in two different formats: `GeoPoint` and `CartesianPoint`. We think that you should define a single `Point` interface, and that all types of points should implement that interface only. We suggest that you pick one coordinate system in which you want to work (either Cartesian or Geodesic) and stick to it and convert any different point to that system **at the moment you retrieve it**. This is related to the global remark about your class hierarchy.

#### EnemyOutdated.java

You should not name a class that way but instead use the `@Deprecated` annotation with the original class name.

#### CacheableUserInfo + DatabaseHelper

* Why are them in the `game` package?
* No JavaDoc
* As mentioned earlier in the semester, it is a huge security risk to store passwords in clear
* An other member of your team used Room to access the SQLite database. We ask you to pick one method and stick to it. We recommend that you use Room, because it produces cleaner and more readable code.

#### Game.java

* Good job for the JavaDoc, try to complete it
* The `Drawable` interface is not a good (nor necessary) abstraction, as it is only implemented by `Game`. It adds noise to the code volume.
* Maybe use the singleton pattern? Most of your methods are static anyway. This allows to easily access the game state from anywhere in the app, and since a player cannot play in multiple games at once it makes sense.
* L63: your method doesn’t work (it looks for a `displayable` in the `updatables` iterable). Your test doesn’t detect it because you test both `updatablesContains` and `displayablesContains` in the same method with the same object. You should not test two different methods in the same test! (Please note that the IDE had highlighted this code! You should make full use of the tools your IDE provides to catch these simple inattention mistakes).
Once you fix the bug, **don’t forget to add a regression test** to make sure it’s not re-introduced again by accident.

#### GameThread.java

* Good JavaDoc
* `FPS`, `targetTime` could be converted to public static final and used as truth everywhere else
* `avgFPS`, `timeMillis` and `waitTime` could be transformed into local variables (to avoid wasting memory space)
* You could find it useful to make `avgFPS` accessible from other classes, if you want them to depend on the actual running speed of the game and not the theoretical one, it’s up to you.

#### DetectableEntity.java

* No JavaDoc
* You shadow the boolean `once` defined in `InteractiveEntity`: this makes it very unclear what is its purpose (and confuses its parent definition)
* Naming: `finalizeEntity` is not a descriptive method name `destroyEntity`, `removeEntity` or `killEntity` would probably be better. And this method should be declared in an `Entity` interface.
* `Sorry it's more complicated due to codeclimate`: you should remove this comment. I’d argue that your code of update is actually clearer now.

#### InteractiveEntity.java

* No Javadoc
* Abstraction problem: a part of your class overlaps with `MovingEntity` (the position field and accessors). You should probably define an abstract class `PositionedEntity`.
* On the other hand, `InteractiveEntity` is only extended by `DetectableEntity`, itself only extended by `ItemBox`. If you don’t plan to create other extensions of these classes, then it means the intermediate abstraction layers can all be merged in `ItemBox`.

#### Inventory.java

* No Javadoc
* Line 36: why not use `inventory.size()` directly?
* Line 6: You should use the type `Map` instead of `LinkedHashMap` in the left-hand part of the declaration (in general, always use the interface type on the left-hand side of declarations).
* Line 32: you should make the map you return immutable by calling `Collections.immutableMap`. The return type of the method should be a simple `Map`.

#### InventoryActivity.java

* Unused `initInventory` method (and it’s only an alias for `initRecyclerView` anyway)
* Typo in word `adpate`

#### Item.java

* No Javadoc
* `name` and `description` should be final
* Otherwise, your item classes and abstraction are really nice

#### ItemBox.java

* No Javadoc
* Lines 33-39: why not take advantage of the `Inventory.addItem` method instead of re-implementing it partially? 

#### ItemViewAdapter.java

* No Javadoc
* `mContext` is never used
* Line 43: you declare a variable with type` LinkedHashMap` without providing the type parameters. This forces you to cast at line 44. This variable should therefore be of type `Map<Item, Integer>`.

#### ItemType.java

* It’s not used, remove it. It’s overlapping with `EntityType.java`.

#### Healthpack.java, Shield.java and Shrinker.java

* These items expose the `getEntityType` method. You should define this method in an interface. (See previous remarks in `EntityType.java` for more details on why).

#### TimedItem.java

Related to [#Enemy.java]: this time, you name the `FPS` magic value. But you should use the now static FPS field of `GameThread.java` - see previous remarks about [#GameThread.java].

#### LeaderboardActivity.java:

* Usually, it’s better to have the ` ViewModel` in the same package as the activity that uses it.
* Typo in word `leaderoardViewModel`

#### GameArea.java

* Nice usage of JavaDoc - just add the Javadoc for the class.
* `radius` and `center` should be set final
* Otherwise very good job

#### GameInfoActivity.java

* The fields should be `private` instead of package-local.
* Line 21: use type `List<T>` and not `ArrayList<T>`

#### RandomGenerator.java

* Line 28: provide a type parameter to the list
* A lot of methods in this class are only used in tests: these methods should be in a specific class in the test sources, or directly in the test class that uses them (if there is only one).
* Should probably be in an `utils` package as it’s very general-purpose.

#### FirebaseAuthentication.java

* Comment lines 22-26 `This regular expression…` applies to what code ?
* Lines 38, 39, 40, 48, 50, 51: this is a very bad practice, you are interacting directly with Activities through static calls but you have no control over the lifecycle of such activities. In fact, you have no guarantees about their existence which could make your app crash non-deterministically. It also poses several other problems: a static public variable is as bad as a global variable (in fact, at bytecode-level it’s identical). This also breaks the separation of concerns, as your `FirebaseAuthentication` class should have absolutely no knowledge about your controllers and views. You should do it the other way around: call the methods from your activities and react at call-site using callbacks or `CompletableFutures`, which will also make independent testing easier.

#### LoginFormActivity.java

* No JavaDoc
* The `public static loggedUser` field could be a local variable: during `onCreate`, you pull the logged in user from the DB if it exists, and then call the appropriate Firebase method to log the user back in. 

#### RegisterFormActivity.java

* Lines 22, 23: these fields should be `private`

#### Displayable.java

* `once()` method has no javadoc and it’s not clear what it is supposed to do.
* The interface itself should have a Javadoc block to explain its purpose.

#### GeoPoint.java

* Good job with the JavaDoc
* Same remarks as before, it would probably easier for you if `GeoPoint` implemented `GenPoint` (or one other solution we suggested to address the coordinates problem).
* This class should be immutable
* Why is there an empty constructor?

#### GoogleMapApi.java

* Line 49: Google Maps informs you when it has detected a position changed with a new position already. Why does `updatePosition` (line 89) redo all the work and refetch the position from a completely different provider? Either use Google Map’s position API or use the device’s location manager only. Also you check the permissions at each update call. You should do it only once at initialization.
* Generally you mix many different things in this class: the location management, but also Google Maps rendering, Bitmap generation. You should split these into appropriate modules (you may already have a suitable class).
* Line 132 and following: it could be more elegant to move a part of the logic to the `Displayable` interface instead of using a giant switch (same change as what was suggested with your Items hierarchy).
* We understand that abstracting Google Maps away is difficult, so don’t lose too much time on this

#### MapApi.java 

* Same problem as before. You expose far too many functionalities. There should be no proprietary type (`GoogleMap`) in the abstraction (only the implementation may have one). Why is there also a `getActivity` here ?

#### Overall remark on the package:

* There is widespread mis-usage of the Singleton pattern in this package. We won’t highlight all of its instances here.
    * In singleton, you have a single static instance, and all methods in the class are instance methods and not static method
    * You did the opposite, with most methods having at least a static version that delegates the call to the singleton instance.
    * At least one of your interfaces (`RemoteToSQLiteAdapter`) declares a `getInstance` method. This doesn’t make sense, as interfaces can only declare instance methods and not static methods. In most cases, when you call `getInstance` you actually already have an instance (otherwise you’d get a `NullPointerException`).
    * **The widespread usage of static fields and singleton patterns makes it very difficult to follow the flow of the code.**
* All the database code is in the package, mixed with UI parts. This makes it difficult to differentiate between the two. I suggest that you move them to a subpackage, or to the main `database` package where other analog classes can already be found. Moreover:
    * `ChatAccessor` should be called `ChatDAO` (by convention)
    * `ChatDatabase` : your app already has a database (`AppDatabase`). In Room, you only define a single `Database` per application. You should therefore move your DAO declaration to the main `db.AppDatabase`.
    * `Chat` should probably be called `Conversation` (would be clearer)
* Lots of commented code: as mentioned earlier, commented code should not be on master. You can use git history to retrieve an old removed piece of code if you end up needing it. In this package in particular, the multiple comments make the code hard to read.
* This package looks like a feature that is still in development. We understand that the Scrum process may force you to push to master unfinished features. However, the parts that are pushed this way should always meet all code quality standards of production code. There is a lot of code that leads to meaningless code (for example : `FriendsListActivity.OnListFragmentInteractionListener`) or implementation stubs.
* **In this package, the code was too confusing for us to be able to make meaningful remarks. We believe it’s as hard for you to understand it as it is for us. You should take the time to rethink the way it works to make it easier to understand.**

#### friends_firestore/

* Your naming should stay consistent. As you had `UserDataController`, this could be a `FriendsController`. Another terminology often used is `Repository`, which I believe is better because the role of a `Repository` is to only fetch data (while a `Controller` can do many things).
* Your `WaitsOnUserFetch` interface is strange: at the same time it is a listener for events that come back from Firebase, but `updateSearch` is a method which is redefined in `RecyclerQueryAdapter` to go back to `RemoteFriendFetcher`. For the first part this is fine but it should be a Listener only. For the second part, why not call `getFriendsFromServer` directly from the call-site?
* Why is your `waiter` stored in `RemoteFriendFetcher`? Shouldn’t it be passed as parameter to `getFriendsFromServer` from your call-site (for instance your Activities) such that you can react directly in your UI when `signalFriendsFetched` is triggered? Here you are setting a mutable field in `RemoteFriendFetcher` as an indirection to do this exact purpose.

#### AddFriendsActivity.java 

* Line 33: you should avoid meaningless logs in production code
* Line 26: why use a list and not a simple reference?
* Several fields are never used
* Line 90: since `container` holds instances of `RecyclerQueryAdapter`, you don’t need to cast them to `WaitsOnUserFetch`. Also, the method to update the search has nothing to do in `WaitsOnUserFetch` in my opinion - this interface should only be used as a callback (see previous remarks).
* Usually in android, it’s the activity that calls the repository and then passes the result of the database calls to the `RecyclerView.Adapter` instance. In your code, it’s the Adapter that listens directly to the database (which forces you to create this bad abstraction discussed earlier). An Adapter should only display the items that are provided to them. 

#### ChatRepository.java

* Should probably be named `SocialRepository` as it also manages friends.

#### ChatActivity.java

* Related to the previous comment about static fields in Activities, no field should be static (**the IDE tells you that it’s a memory leak**).
* There shouldn’t be so many public methods: as said previously, you have no guarantee outside the Activity class over its existence.
* Interesting usage of the `MessageDecorator`, the class could be package-private.

#### RecyclerQueryAdapter.java

* Same remark as for [#AddFriendsActivity.java]: adapters should not be responsible for pulling the data but merely transforming the list to a displayable view.

#### WaitOn*

* You should use genericity to declare a single `WaitOn<T>` class (or `Callback<T>`) instead of writing 3 classes declaring almost the same thing.

## Testing

* Your testing code should be as readable as your production code
* All tests should meaningfully assert conditions or check useful properties

## Conclusion
You had nice some ideas for abstracting and organizing your code. Unfortunately, it seems your implementation is lacking: you should be more careful with thinking about the technical architecture of your code before writing it. Abstraction and design patterns are not always a silver bullet against complexity, and they may also worsen it. Try to be more precise, organized and rigorous in your design process, and focus on readability and reducing the cognitive load.

We think that a lot of abstraction/class hierarchy problems may become clearer for you (and easier to solve) once you’ll start to focus on having a fully playable game. You should probably first work on this, as your other features will be moot should the game not work. We expect you to improve at least the general architecture of your code, as it unfortunately doesn’t match the quality standards of the SDP course as of now.

Good luck and keep up the work.
